docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md

```diff
@@ -0,0 +1,83 @@
+---
+title: Adding PMD support for a new ANTLR grammar based language
+short_title: Adding a new language with ANTLR
+tags: [devdocs, extending]
+summary: "How to add a new language to PMD using ANTLR grammar."
+last_updated: July 21, 2019
+sidebar: pmd_sidebar
+permalink: pmd_devdocs_major_adding_new_language.html
+folder: pmd/devdocs
+---
+
+
+## 1.  Start with a new sub-module.
+*   See pmd-swift for examples.
+
+## 2.  Implement an AST parser for your language
+*	ANTLR gives you this for free.
+
+## 3.  Create AST node classes
+*	We provide an [AntlrBaseNode](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/AntlrBaseNode.java). 
+*	We override ANTLR auto-generated code to provide this for free, you need to add an ANT script similar to the [swift scrip](https://github.com/pmd/pmd/blob/master/pmd-swift/src/main/ant/antlr4.xml).
+*	You can extend AntlrBaseNode and override any method that you require, but on most cases you won't need to do anything.
+
+## 4.  Compile your parser 
+*	We override ANTLR auto-generated code to provide this for free, similar to the step before, you will need to use the ANT script.
+*	You should review the [swift pom](https://github.com/pmd/pmd/blob/master/pmd-swift/pom.xml). Don't forget to enable visitor generation property. 
+
+## 5.  Create a TokenManager
+*   We provide a default implementation using [AntlrTokenManager](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrTokenManager.java) that uses an [AntlrTokenizer](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/AntlrTokenizer.java). 
+*	You must create your own AntlrTokenizer such as we do with [SwiftTokenizer](https://github.com/pmd/pmd/blob/master/pmd-swift/src/main/java/net/sourceforge/pmd/cpd/SwiftTokenizer.java).
+*	If you wish to filter specific tokens you can create your own implementation of [BaseTokenFilter](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/token/internal/BaseTokenFilter.java) as we did with [SwiftTokenFilter](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/token/AntlrTokenFilter.java).
```

Comment:
if we don't want to, what do we do?

Reply:
I've just updated this scenario, pointing to AntlrTokenFilter, the default implementation we provide